Add information about implicit version to AspNetCore references (#11736)
dotnet/docs: docs/core/tools/csproj.md

## Implicit package references

> Metapackages are implicitly referenced based on the target framework(s) specified in the `<TargetFramework>` or `<TargetFrameworks>` property of your project file. `<TargetFrameworks>` is ignored if `<TargetFramework>` is specified, independent of order. For more information, see [Packages, metapackages and frameworks](/dotnet/core/packages). 

Reviewer: Let's use relative links to get validation at build time 

```suggestion
Metapackages are implicitly referenced based on the target framework(s) specified in the `<TargetFramework>` or `<TargetFrameworks>` property of your project file. `<TargetFrameworks>` is ignored if `<TargetFramework>` is specified, independent of order. For more information, see [Packages, metapackages and frameworks](../packages.md). 
```

* When targeting .NET Core or .NET Standard, never have an explicit reference to the `Microsoft.NETCore.App` or `NETStandard.Library` metapackages via a `<PackageReference>` item in your project file.
* If you need a specific version of the runtime when targeting .NET Core, you should use the `<RuntimeFrameworkVersion>` property in your project (for example, `1.0.4`) instead of referencing the metapackage.
  * This might happen if you are using [self-contained deployments](../deploying/index.md#self-contained-deployments-scd) and you need a specific patch version of 1.0.0 LTS runtime, for example.
* If you need a specific version of the `NETStandard.Library` metapackage when targeting .NET Standard, you can use the `<NetStandardImplicitPackageVersion>` property and set the version you need.
* Don't explicitly add or update references to either the `Microsoft.NETCore.App` or `NETStandard.Library` metapackage in .NET Framework projects. If any version of `NETStandard.Library` is needed when using a .NET Standard-based NuGet package, NuGet automatically installs that version.

## Implicit version for some package references

> Most usages of [`<PackageReference>`](#packagereference) require setting the `Version` attribute to specify the NuGet package version to be used. When referencing [Microsoft.AspNetCore.App](/aspnet/core/fundamentals/metapackage-app) or [Microsoft.AspNetCore.All](/aspnet/core/fundamentals/metapackage), however, the  attribute is unnecessary. The .NET Core SDK can automatically select the version of these packages which should be used.

Reviewer: ```suggestion
Most usages of [`<PackageReference>`](#packagereference) require setting the `Version` attribute to specify the NuGet package version to be used. When referencing [Microsoft.AspNetCore.App](/aspnet/core/fundamentals/metapackage-app) or [Microsoft.AspNetCore.All](/aspnet/core/fundamentals/metapackage), however, the  attribute is unnecessary. The .NET Core SDK can automatically select the version of these packages that should be used.
```

> Most usages of [`<PackageReference>`](#packagereference) require setting the `Version` attribute to specify the NuGet package version to be used. When referencing [Microsoft.AspNetCore.App](/aspnet/core/fundamentals/metapackage-app) or [Microsoft.AspNetCore.All](/aspnet/core/fundamentals/metapackage), however, the  attribute is unnecessary. The .NET Core SDK can automatically select the version of these packages which should be used.

Reviewer 2: I think this should mention that the version is only automatically selected when targeting .NET Core 2.1 and 2.2.  If you're targeting 2.0 or earlier, you still need to specify the version.

Also the implicit version only worked with the Web SDK in the 2.1.xxx SDKs, so you need to be using a 2.2.xxx .NET Core SDK for the implicit versioning to work better, even if you are targeting .NET Core 2.1.

### Recommendation

> When referencing Microsoft.AspNetCore.App or .All, do not specify the version of this package. If a version is specified, the SDK may produce warning NETSDK1071. To fix this warning, the recommended usage is:

Reviewer: ```suggestion
When referencing the `Microsoft.AspNetCore.App` or `Microsoft.AspNetCore.All` packages, do not specify their version. If a version is specified, the SDK may produce warning NETSDK1071. To fix this warning, remove the package version like in the following example:
```

```xml
<ItemGroup>
  <PackageReference Include="Microsoft.AspNetCore.App" />
</ItemGroup>
```

> These references to ASP.NET Core metapackages have a slightly different behavior from most normal NuGet packages. Applications that use these metapackages automatically take advantage of the ASP.NET Core shared framework. When you use the  metapackages, **no** assets from the referenced ASP.NET Core NuGet packages are deployed with the application&mdash;the ASP.NET Core shared framework contains these assets. The assets in the shared framework are precompiled to improve application startup time. For more information, see "shared framework" in [.NET Core distribution packaging](/dotnet/core/build/distribution-packaging).

Reviewer 3: > precompiled

Kind of an odd word. It's all DLLs, so they're already "compiled". Maybe better to say "are optimized for the target platform"?

> These references to ASP.NET Core metapackages have a slightly different behavior from most normal NuGet packages. Applications that use these metapackages automatically take advantage of the ASP.NET Core shared framework. When you use the  metapackages, **no** assets from the referenced ASP.NET Core NuGet packages are deployed with the application&mdash;the ASP.NET Core shared framework contains these assets. The assets in the shared framework are optimized for the target platform to improve application startup time. For more information, see "shared framework" in [.NET Core distribution packaging](/dotnet/core/build/distribution-packaging).

Reviewer: ```suggestion
These references to ASP.NET Core metapackages have a slightly different behavior from most normal NuGet packages. Applications that use these metapackages automatically take advantage of the ASP.NET Core shared framework. When you use the metapackages, **no** assets from the referenced ASP.NET Core NuGet packages are deployed with the application—the ASP.NET Core shared framework contains these assets. The assets in the shared framework are optimized for the target platform to improve application startup time. For more information about shared framework, see [.NET Core distribution packaging](../build/distribution-packaging.md).
```

> These references to ASP.NET Core metapackages have a slightly different behavior from most normal NuGet packages. Applications that use these metapackages automatically take advantage of the ASP.NET Core shared framework. When you use the  metapackages, **no** assets from the referenced ASP.NET Core NuGet packages are deployed with the application&mdash;the ASP.NET Core shared framework contains these assets. The assets in the shared framework are optimized for the target platform to improve application startup time. For more information, see "shared framework" in [.NET Core distribution packaging](/dotnet/core/build/distribution-packaging).

Reviewer 2: I would say "Framework dependent applications that use these metapackages automatically...", because self-contained apps do end up with the implementations for the package's APIs.

> If a version *is* specified, it is treated as the *minimum* version of ASP.NET Core shared framework for framework-dependent deployments and as a *maximum* version for self-contained deployments. This can have the following consequences:

Reviewer: ```suggestion
If a version *is* specified, it's treated as the *minimum* version of ASP.NET Core shared framework for framework-dependent deployments and as a *maximum* version for self-contained deployments. This can have the following consequences:
```

> If a version *is* specified, it is treated as the *minimum* version of ASP.NET Core shared framework for framework-dependent deployments and as a *maximum* version for self-contained deployments. This can have the following consequences:

Reviewer 2: Isn't it treated as the *exact* version for self-contained deployments, not a maximum in that case?

> * If the version of ASP.NET Core installed on the server is less than the version specified on the PackageReference, the .NET Core process will fail to launch. Updates to the metapackage are often available on NuGet.org before updates have been made available in hosting environments such as Azure. Updating the version on the PackageReference to ASP.NET Core could cause an application deployment failure.

Reviewer 3: > could cause an application deployment failure

Maybe: "could cause a deployed application to fail"

That is, the deployment will work just fine. The app won't run, though.

> * If the version of ASP.NET Core installed on the server is less than the version specified on the PackageReference, the .NET Core process will fail to launch. Updates to the metapackage are often available on NuGet.org before updates have been made available in hosting environments such as Azure. Updating the version on the PackageReference to ASP.NET Core could cause a deployed application to fail.

Reviewer: ```suggestion

* If the version of ASP.NET Core installed on the server is less than the version specified on the PackageReference, the .NET Core process fails to launch. Updates to the metapackage are often available on NuGet.org before updates have been made available in hosting environments such as Azure. Updating the version on the PackageReference to ASP.NET Core could cause a deployed application to fail.
```

> * If the application is deployed as a [self-contained deployment](../deploying/index.md#self-contained-deployments-scd), the application may not contain the latest security updates to .NET Core. When a version is not specified, the SDK can automatically include the newest version of ASP.NET Core in the self-contained deployment.

Reviewer: ```suggestion
* If the application is deployed as a [self-contained deployment](../deploying/index.md#self-contained-deployments-scd), the application may not contain the latest security updates to .NET Core. When a version isn't specified, the SDK can automatically include the newest version of ASP.NET Core in the self-contained deployment.
```
